src/linux_udev.rs

```diff
@@ -0,0 +1,762 @@
+        if let Some((usage_page, usage)) = usage_pair {
+            return Some((usage_page, usage, advanced));
+        }
+    }
+
+    if let (true, Some(usage)) = (initial, usage) {
+        return Some((usage_page, usage, advanced));
+    }
+
+    None
+}
+
+/// Gets the size of the HID item at the given position
+///
+/// Returns data_len and key_size when successful
+fn hid_item_size(desc: &[u8]) -> Option<(usize, usize)> {
+    let key = desc[0];
+
+    // Long Item. Next byte contains the length of the data section.
+    if (key & 0xf0) == 0xf0 {
+        if !desc.is_empty() {
+            return Some((desc[1].into(), 3));
+        }
+
+        // Malformed report
+        return None;
+    }
+
+    // Short Item. Bottom two bits contains the size code
+    match key & 0x03 {
+        v @ 0 | v @ 1 | v @ 2 => Some((v.into(), 1)),
+        3 => Some((4, 1)),
+        _ => unreachable!(), // & 0x03 means this can't happen
+    }
+}
+
+/// Get the bytes from a HID report descriptor
+///
+/// Must only be called with `num_bytes` 0, 1, 2 or 4.
+fn hid_report_bytes(desc: &[u8], num_bytes: usize) -> u32 {
+    if num_bytes >= desc.len() {
+        return 0;
+    }
+
+    match num_bytes {
+        0 => 0,
+        1 => desc[1] as u32,
+        2 => {
+            let bytes = [desc[1], desc[2], 0, 0];
+            u32::from_le_bytes(bytes)
+        }
+        4 => u32::from_le_bytes(desc[1..=4].try_into().unwrap()),
+        _ => unreachable!(),
+    }
+}
+
+/// Get the attribute from the device and convert it into a [`WcharString`].
+fn attribute_as_wchar(dev: &udev::Device, attr: &str) -> WcharString {
+    dev.attribute_value(attr)
+        .map(Into::into)
+        .map(osstring_to_string)
+        .unwrap_or(WcharString::None)
+}
+
+/// Get the attribute from the device and convert it into a i32
+///
+/// On error or if the attribute is not found, it returns -1;
+fn attribute_as_i32(dev: &udev::Device, attr: &str) -> i32 {
+    dev.attribute_value(attr)
+        .and_then(OsStr::to_str)
+        .and_then(|v| i32::from_str_radix(v, 16).ok())
+        .unwrap_or(-1)
+}
+
+/// Get the attribute from the device and convert it into a u16
+///
+/// On error or if the attribute is not found, it returns 0;
+fn attribute_as_u16(dev: &udev::Device, attr: &str) -> u16 {
+    dev.attribute_value(attr)
+        .and_then(OsStr::to_str)
+        .and_then(|v| u16::from_str_radix(v, 16).ok())
+        .unwrap_or(0)
+}
+
+/// Convert a [`OsString`] into a [`WcharString`]
+fn osstring_to_string(s: OsString) -> WcharString {
+    match s.into_string() {
+        Ok(s) => WcharString::String(s),
+        Err(_) => panic!("udev strings should always be utf8"),
+    }
+}
+
+/// Parse a HID_ID string to find the bus type, the vendor and product id
+///
+/// These strings would be of the format
+///     type vendor   product
+///     0003:000005AC:00008242
+fn parse_hid_vid_pid(s: &str) -> Option<(u16, u16, u16)> {
```

Comment:
Probably a good idea to out parsing related functionality (which might even be shared between backends) into a different file.
Also the unit tests for parsing could be moved to that file.

Reply:
Looking at what C hidapi does for macOS, it doesn't look like we'd need them for anything else. The vendor and product ids are parsed by the system and given to us, as are the usage_page and usage information. Same goes with Windows.

So we can move them somewhere else if you prefer, like with the ioctls, but it doesn't look like it's something we will be re-using.